aws/resource_aws_eks_addon.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	if v, ok := d.GetOk("service_account_role_arn"); ok {
+		input.ServiceAccountRoleArn = aws.String(v.(string))
+	}
+
+	if v := d.Get("tags").(map[string]interface{}); len(v) > 0 {
+		input.Tags = keyvaluetags.New(v).IgnoreAws().EksTags()
+	}
+
+	log.Printf("[DEBUG] Creating EKS Cluster Addon %s", addonName)
+
+	err := resource.RetryContext(ctx, 1*time.Minute, func() *resource.RetryError {
+		_, err := conn.CreateAddonWithContext(ctx, input)
+		if err != nil {
+			//
+			if isAWSErr(err, eks.ErrCodeInvalidParameterException, "CREATE_FAILED") {
+				return resource.RetryableError(err)
+			}
+			if isAWSErr(err, eks.ErrCodeInvalidParameterException, "does not exist") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if isResourceTimeoutError(err) {
+		_, err = conn.CreateAddonWithContext(ctx, input)
+	}
+	if err != nil {
+		return diag.Errorf("error creating EKS Cluster Addon (%s): %s", addonName, err)
+	}
+
+	d.SetId(addonName)
+
+	stateConf := resource.StateChangeConf{
+		Pending: []string{eks.AddonStatusCreating},
+		Target:  []string{eks.AddonStatusActive},
+		Timeout: d.Timeout(schema.TimeoutCreate),
+		Refresh: refreshEksAddonStatusContext(ctx, conn, addonName, clusterName),
+	}
+	// TODO: Handle unexpected sate CREATE_FAILED and retry?
+	// Error: unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: %!s(<nil>)
+	// How to reproduce: taint eks addon resource and try to do terraform apply
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceAwsEksAddonRead(ctx, d, meta)
+}
+
+func resourceAwsEksAddonRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).eksconn
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	addonName := d.Get("addon_name").(string)
+	clusterName := d.Get("cluster_name").(string)
+
+	input := &eks.DescribeAddonInput{
+		AddonName:   aws.String(addonName),
+		ClusterName: aws.String(clusterName),
+	}
+
+	log.Printf("[DEBUG] Reading EKS Cluster Addon: %s", input)
+	output, err := conn.DescribeAddonWithContext(ctx, input)
+	if err != nil {
+		if isAWSErr(err, eks.ErrCodeResourceNotFoundException, "") {
+			log.Printf("[WARN] EKS Cluster Addon (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("error reading EKS Cluster (%s): %s", d.Id(), err)
+	}
+
+	addon := output.Addon
+	if addon == nil {
+		log.Printf("[WARN] EKS Cluster (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	d.SetId(addonName)
+	d.Set("arn", addon.AddonArn)
+	d.Set("status", addon.Status)
+	d.Set("addon_version", addon.AddonVersion)
+	d.Set("service_account_role_arn", addon.ServiceAccountRoleArn)
+	d.Set("created_at", aws.TimeValue(addon.CreatedAt).String())
+	d.Set("modified_at", aws.TimeValue(addon.ModifiedAt).String())
```

Comment:
I think you need to define `status`, `created_at` and `modified_at` field to the schema.

Reply:
Yes for sure. I was unsure if we actually need these attributes initially.
Looking at some other AWS resources (resource_aws_eks_cluster for example) I see that all of these definitely should be added.